File: Elementary/panindrome/125.ValidPalindrome.java
package elementary.panindrome;

class Solution {
    /** two pointers */
    public boolean isPalindrome(String s) {
        int n = s.length();
        // String s = s.toUpperCase(); // alternatively
        for (int i = 0, j = n - 1; i < j; i++, j--) {
            while (i < j && !Character.isLetterOrDigit(s.charAt(i)))
                i++;
            while (i < j && !Character.isLetterOrDigit(s.charAt(j)))
                j--;
            if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j)))
                return false;
        }
        return true;
    }

    /** compar with reverse */
    public boolean isPalindrome2(String s) {
        StringBuilder sb = new StringBuilder("");
        for (char c : s.toCharArray()) {
            if (Character.isLetterOrDigit(c))
                sb.append(Character.toLowerCase(c));
        }
        String s1 = sb.toString();
        String s2 = sb.reverse().toString();
        return s1.equals(s2);
    }
}

File: Elementary/palindrome/125.ValidPalindrome.cpp
class Solution {
public:
    // two-pointers
    bool isPalindrome(string s) {
        string new_s = "";
        for (char c : s) {
            if (isalnum(c))
                new_s += tolower(c);
        }
        if (new_s.size() < 2) return true;
        for (int i = 0, j = new_s.size() - 1; i < j; i++, j--) {
            if (new_s[i] != new_s[j]) return false;
        }
        return true;
    }

    // in-place
    bool isPalindrome2(string s) {
        for (int i = 0, j = s.size() - 1; i < j; i++, j--) {
            while (i < j && !isalnum(s[i]))
                i++;
            while (i < j && !isalnum(s[j]))
                j--;
            if (tolower(s[i]) != tolower(s[j])) return false;
        }
        return true;
    }
};